dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/TemplateAPIImpl.java

```diff
@@ -258,43 +258,51 @@ public Template saveTemplate(Template template, Host destination, User user, boo
 		template.setModDate(new Date());
 		template.setModUser(user.getUserId());
 
-		//we need to replace older container parse syntax with updated syntax
-		updateParseContainerSyntax(template);
+
 
 		//it saves or updates the asset
-		if(existingInode)
+		if(existingInode) {
 		    // support for existing inode
 		    save(template,template.getInode());
-		else
+		}else {
 		    save(template);
-
+		}
 		APILocator.getVersionableAPI().setWorking(template);
 
 
-		///parses the body tag to get all identifier ids and saves them as children
-		List<Container> containers = getContainersInTemplate(template, user, respectFrontendRoles);
-		
-
-        //Adding the permissions for this Permissionable to cache
-        permissionAPI.addPermissionsToCache( template );
         return template;
 	}
 
-	@CloseDBIfOpened
-	@Override
-	public List<Container> getContainersInTemplate(Template template, User user, boolean respectFrontendRoles) throws DotDataException, DotSecurityException {
-	    if(template!=null && InodeUtils.isSet(template.getInode())) {
+    @CloseDBIfOpened
+    @Override
+    public List<Container> getContainersInTemplate(Template template, User user, boolean respectFrontendRoles)
+            throws DotDataException, DotSecurityException {
 
-    		if (!permissionAPI.doesUserHavePermission(template, PermissionAPI.PERMISSION_READ, user, respectFrontendRoles)) {
-    			throw new DotSecurityException("You don't have permission to read the source file.");
-    		}
+        // 100 pages should be more than enough to get all the containers on a given template
+        // Trying to avoid the case where 10000 pages might use the same template
+        List<Contentlet> pages = APILocator.getHTMLPageAssetAPI().findPagesByTemplate(template, user, respectFrontendRoles, 100);
 
-    		return templateFactory.getContainersInTemplate(template, user, respectFrontendRoles);
-	    }
-	    else {
-	        return null;
-	    }
-	}
+        List<Container> containers = new ArrayList<>();
+
+        for (Contentlet page : pages) {
+            Set<String> containerId =
+                    MultiTreeFactory.getMultiTrees(page.getIdentifier())
+                    .stream()
+                    .map(mtree -> mtree.getContainer())
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Specify a type for: 'mtree' [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2211)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this lambda with a method reference. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)